Mathlib/LinearAlgebra/TensorProduct/Tower.lean

```diff
@@ -176,27 +183,210 @@ nonrec def mk : M →ₗ[A] N →ₗ[R] M ⊗[R] N :=
+    (congr f g).symm (p ⊗ₜ q) = f.symm p ⊗ₜ g.symm q :=
+  rfl
+
+end Semiring
+
+section CommSemiring
+
+variable [CommSemiring R] [CommSemiring A] [Semiring B] [Algebra R A] [Algebra R B]
+
+variable [AddCommMonoid M] [Module R M] [Module A M] [Module B M]
+variable [IsScalarTower R A M] [IsScalarTower R B M] [SMulCommClass A B M]
+
+variable [AddCommMonoid N] [Module R N]
+
+variable [AddCommMonoid P] [Module R P] [Module A P] [Module B P]
+variable [IsScalarTower R A P] [IsScalarTower R B P] [SMulCommClass A B P]
+
+variable [AddCommMonoid Q] [Module R Q]
+
+variable (R A B M N P Q)
+
 attribute [local ext high] TensorProduct.ext
 
+section assoc
+variable [Algebra A B] [IsScalarTower A B M]
+
 /-- Heterobasic version of `TensorProduct.assoc`:
 
 Linear equivalence between `(M ⊗[A] N) ⊗[R] P` and `M ⊗[A] (N ⊗[R] P)`. -/
-def assoc : (M ⊗[A] P) ⊗[R] N ≃ₗ[A] M ⊗[A] P ⊗[R] N :=
+def assoc : (M ⊗[A] P) ⊗[R] Q ≃ₗ[B] M ⊗[A] (P ⊗[R] Q) :=
   LinearEquiv.ofLinear
-    (lift <|
-      TensorProduct.uncurry A _ _ _ <| comp (lcurry R A _ _ _) <| TensorProduct.mk A M (P ⊗[R] N))
-    (TensorProduct.uncurry A _ _ _ <|
-      comp (uncurry R A _ _ _) <| by
-        apply TensorProduct.curry
-        exact mk R A _ _)
+    (lift <| lift <| lcurry R A B P Q _ ∘ₗ mk A B M (P ⊗[R] Q))
+    (lift <| uncurry R A B P Q _ ∘ₗ curry (mk R B _ Q))
+    (by ext; rfl)
+    (by ext; rfl)
+-- porting note: new `B` argument
+#align tensor_product.algebra_tensor_module.assoc TensorProduct.AlgebraTensorModule.assocₓ
+
+variable {M P N Q}
+
+@[simp]
+theorem assoc_tmul (m : M) (p : P) (q : Q) :
+    assoc R A B M P Q ((m ⊗ₜ p) ⊗ₜ q) = m ⊗ₜ (p ⊗ₜ q) :=
+  rfl
+
+@[simp]
+theorem assoc_symm_tmul (m : M) (p : P) (q : Q) :
+    (assoc R A B M P Q).symm (m ⊗ₜ (p ⊗ₜ q)) = (m ⊗ₜ p) ⊗ₜ q :=
+  rfl
+
+end assoc
+
+section leftComm
+
+/-- Heterobasic version of `TensorProduct.leftComm` -/
+def leftComm : M ⊗[A] (P ⊗[R] Q) ≃ₗ[A] P ⊗[A] (M ⊗[R] Q) :=
+  let e₁ := (assoc R A A M P Q).symm
+  let e₂ := congr (TensorProduct.comm A M P) (1 : Q ≃ₗ[R] Q)
+  let e₃ := (assoc R A A P M Q)
+  e₁ ≪≫ₗ (e₂ ≪≫ₗ e₃)
+
+variable {M N P Q}
+
+@[simp]
+theorem leftComm_tmul (m : M) (p : P) (q : Q) :
+    leftComm R A M P Q (m ⊗ₜ (p ⊗ₜ q)) = p ⊗ₜ (m ⊗ₜ q) :=
+  rfl
+
+@[simp]
+theorem leftComm_symm_tmul (m : M) (p : P) (q : Q):
+    (leftComm R A M P Q).symm (p ⊗ₜ (m ⊗ₜ q)) = m ⊗ₜ (p ⊗ₜ q) :=
+  rfl
+
+end leftComm
+
+section rightComm
+
+/-- A tensor product analogue of `mul_right_comm`. -/
+def rightComm : (M ⊗[A] P) ⊗[R] Q ≃ₗ[A] (M ⊗[R] Q) ⊗[A] P :=
+  LinearEquiv.ofLinear
+    (lift <| TensorProduct.lift <| LinearMap.flip <|
+      lcurry R A A M Q ((M ⊗[R] Q) ⊗[A] P) ∘ₗ (mk A A (M ⊗[R] Q) P).flip)
+    (TensorProduct.lift <| lift <| LinearMap.flip <|
+      (TensorProduct.lcurry A M P ((M ⊗[A] P) ⊗[R] Q)).restrictScalars R
+        ∘ₗ (mk R A (M ⊗[A] P) Q).flip)
     (by
-      ext
-      rfl)
+      refine (TensorProduct.ext <| ext <| fun m q => LinearMap.ext <| fun p => ?_)
```

Comment:
You can make that `?_` into a `rfl`, but if you try it with the one two lines down you get a timeout.

Reply:
I went for a compromise here and below.